src/read/wasm.rs

```diff
@@ -54,23 +63,229 @@ impl<'data> WasmFile<'data> {
+                            flags: SymbolFlags::None,
+                        });
+                    }
+                }
+                wp::SectionCode::Function => {
+                    local_func_kinds = vec![
+                        LocalFunctionKind::Unknown;
+                        section
+                            .get_function_section_reader()
+                            .map_err(|_| "Couldn't read header of the function section")?
+                            .get_count() as usize
+                    ];
+                }
+                wp::SectionCode::Export => {
+                    if let Some(main_file_symbol) = main_file_symbol.take() {
+                        file.symbols.push(main_file_symbol);
+                    }
+
+                    for export in section
+                        .get_export_section_reader()
+                        .map_err(|_| "Couldn't read header of the export section")?
+                    {
+                        let export = export.map_err(|_| "Couldn't read an export item")?;
+
+                        file.symbols.push(Symbol {
+                            name: Some(export.field),
+                            address: 0,
+                            size: 0,
+                            kind: match export.kind {
+                                wp::ExternalKind::Function => {
+                                    if let Some(local_func_id) =
+                                        export.index.checked_sub(imported_funcs_count)
+                                    {
+                                        let local_func_kind =
+                                            &mut local_func_kinds[local_func_id as usize];
+                                        if let LocalFunctionKind::Unknown = local_func_kind {
+                                            *local_func_kind = LocalFunctionKind::Exported {
+                                                symbol_ids: Vec::new(),
+                                            };
+                                        }
+                                        let symbol_ids = match local_func_kind {
+                                            LocalFunctionKind::Exported { symbol_ids } => {
+                                                symbol_ids
+                                            }
+                                            _ => unreachable!(),
+                                        };
+                                        symbol_ids.push(file.symbols.len() as u32);
+                                    }
+                                    SymbolKind::Text
+                                }
+                                wp::ExternalKind::Table
+                                | wp::ExternalKind::Memory
+                                | wp::ExternalKind::Global => SymbolKind::Data,
+                            },
+                            section: SymbolSection::Section(SectionIndex(SECTION_EXPORT)),
```

Comment:
This doesn't seem right to me. This field is meant to tell you which section to look in to get the data for the symbol. Are the instructions used to define the function in the export section or the code section?

Reply:
This is, unfortunately, a bit convoluted due to the nature of streaming parsing, but the idea is that when we read this section, we start with all symbols as just "export" symbols because we don't know yet where in the file they're pointing to.

Later, when we read the "code" section and find out that this export symbol was pointing to a local function, we update its `section`, `kind`, `address` and `size` to mark it as such (see line 237).

Otherwise, it remains an "export" section symbol.

Reply:
Let's clear up the above comment before discussing this one further, because I think this has the same misconception about what the `section` field is for.

Reply:
Like above, I believe this is resolved now.